src/backchannel_logout.rs

```diff
@@ -0,0 +1,469 @@
+//! This module implements components needed for [back-channel logout]
+//!
+//! [back-channel logout]: <https://openid.net/specs/openid-connect-backchannel-1_0.html>
+
+use std::collections::HashMap;
+
+use chrono::{DateTime, Utc};
+use serde::{de::Error, Deserialize};
+
+use crate::{
+    types::helpers::{deserialize_string_or_vec, serde_utc_seconds},
+    types::SessionIdentifier,
+    Audience, IssuerUrl, SubjectIdentifier,
+};
+
+/// The Logout Token as defined in [section 2.4] of the [OpenID Connect Back-Channel Logout spec][1]
+///
+/// [section 2.4]: <https://openid.net/specs/openid-connect-backchannel-1_0.html#LogoutToken>
+/// [1]: <https://openid.net/specs/openid-connect-backchannel-1_0.html>
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct LogoutToken {
+    /// The issuer of this token
+    iss: IssuerUrl,
+    /// The audience this token is intended for
+    aud: Vec<Audience>,
+    /// Time at which this token was issued
+    iat: DateTime<Utc>,
+    /// The unique identifier for this token. This can be used to detect
+    /// replay attacks.
+    jti: String,
+    identifier: Identifier,
+    events: HashMap<String, serde_json::Value>,
+}
+
+impl LogoutToken {
+    /// The `iss` claim
+    pub fn issuer(&self) -> &IssuerUrl {
+        &self.iss
+    }
+
+    /// The `aud` claim
+    pub fn audiences(&self) -> impl Iterator<Item = &Audience> {
+        self.aud.iter()
+    }
+
+    /// The `iat` claim
+    pub fn issue_time(&self) -> DateTime<Utc> {
+        self.iat
+    }
+
+    /// The `jti` claim. It's the unique identifier for this token and can be
+    /// used to detect replay attacks.
+    pub fn jti(&self) -> &str {
+        &self.jti
+    }
+
+    /// As per spec, a [`LogoutToken`] MUST either have the `sub`  or `sid`
+    /// claim and MAY contain both. You can match the [`Identifier`] to detect
+    /// which claims are present.
+    pub fn identifier(&self) -> &Identifier {
+        &self.identifier
+    }
+
+    /// A [`LogoutToken`] is compatible with the [SET standard from RFC 8417][1]
+    ///
+    /// [1]: <https://www.rfc-editor.org/info/rfc8417>
+    pub fn events(&self) -> &HashMap<String, serde_json::Value> {
+        &self.events
+    }
+}
+impl<'de> Deserialize<'de> for LogoutToken {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let value: serde_json::Value = serde_json::Value::deserialize(deserializer)?;
+        if let serde_json::Value::Object(ref map) = value {
+            if map.contains_key("nonce") {
+                return Err(<D::Error as Error>::custom("nonce claim is prohibited"));
+            }
+        }
+
+        #[derive(Deserialize)]
+        struct Repr {
```

Comment:
Following the suggestions above, I think the duplicate `Repr` and custom `Deserialize` implementation can probably be removed in favor of serde annotations where needed